### PR TITLE
[DOC][C API][DYNAMIC SHAPE] Add doc for C API dynamic shape

### PR DESCRIPTION
**Pull request by zhaixuejun1993**
### Details:
 - add descriptions for C API dynamic shape

### Tickets:
- 93265

### CODE REVIEW

**Comment by ilyachur** [comment]
Am I right that this PR depends on https://github.com/openvinotoolkit/openvino/pull/13290?

**Comment by riverlijunjie** [comment]
> Am I right that this PR depends on #13290?



> Am I right that this PR depends on #13290?

yes, #13290 is for basic pipeline table, and this PR is for dynamic shape documentation.

File: docs/snippets/ov_dynamic_shapes.c [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+ov_model_t* model = NULL;
+ov_core_read_model(core, "model.xml", NULL, &model);
+
+// Set one static dimension (= 1) and another dynamic dimension (= Dimension())
+{
+ov_partial_shape_t partial_shape;
+int64_t rank = 2;
+ov_dimension_t dims[2] = {{1, 1}, {-1, -1}};
+ov_partial_shape_create(rank, dims, &partial_shape);
+ov_model_reshape(model, "tensor_name", &partial_shape, 1); // {1,?}
```
Comment (riverlijunjie) [inline docs/snippets/ov_dynamic_shapes.c:21]: Real dynamic shape support should not call reshape interface, but GPU plugin doesn't enable real dynamic shape so far.

File: docs/snippets/ov_dynamic_shapes.c [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+ov_core_read_model(core, "model.xml", NULL, &model);
+//! [ov_dynamic_shapes:print_dynamic]
+// Print output partial shape
+ov_output_port_t* output_port = NULL;
+ov_model_output(model, &output_port);
+ov_partial_shape_t partial_shape;
+ov_port_get_partial_shape(output_port, &partial_shape);
+char * str_partial_shape = ov_partial_shape_to_string(partial_shape);
+printf("The output partial shape: %s", str_partial_shape);
+
```
Comment (riverlijunjie) [inline docs/snippets/ov_dynamic_shapes.c:68]: We need free str_partial_shape here.

File: docs/snippets/ov_dynamic_shapes.c [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+ov_model_reshape(model, "tensor_name", &partial_shape, 1); // {1,?}
+}
+
+// Or set both dimensions as dynamic if both are going to be changed dynamically
+{
+ov_partial_shape_t partial_shape;
+int64_t rank = 2;
+ov_dimension_t dims[2] = {{-1, -1}, {-1, -1}};
+ov_partial_shape_create(rank, dims, &partial_shape);
+ov_model_reshape(model, "tensor_name", &partial_shape, 1); // {?,?}
```
Comment (riverlijunjie) [inline docs/snippets/ov_dynamic_shapes.c:30]: For C sample, we need release all allocated objects explicitly - call ov_partial_shape_free to release partial_shape.

File: docs/snippets/ov_dynamic_shapes.c [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+char * str_partial_shape = ov_partial_shape_to_string(partial_shape);
+printf("The output partial shape: %s", str_partial_shape);
+
+// Print input partial shape
+ov_output_port_t* input_port = NULL;
+ov_model_input(model, &input_port);
+ov_partial_shape_t partial_shape;
+ov_port_get_partial_shape(input_port, &partial_shape);
+char * str_partial_shape = ov_partial_shape_to_string(partial_shape);
+printf("The input partial shape: %s", str_partial_shape);
```
Comment (riverlijunjie) [inline docs/snippets/ov_dynamic_shapes.c:75]: Do clean up here for C sample.

File: docs/snippets/ov_dynamic_shapes.c [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+
+//! [ov_dynamic_shapes:detect_dynamic]
+ov_model_t* model = NULL;
+ov_core_read_model(core, "model.xml", NULL, &model);
+
+ov_output_port_t* input_port = NULL;
+ov_model_input_by_index(model, 0, &input_port);
+ov_partial_shape_t partial_shape;
+ov_port_get_partial_shape(input_port, &partial_shape);
+if (ov_partial_shape_is_dynamic(partial_shape)) {
```
Comment (riverlijunjie) [inline docs/snippets/ov_dynamic_shapes.c:91]: It should follow C++ snippet code for dynamic shape. 
Will be possible add below api to check whether model is dynamic?
`bool ov_model_is_dynamic(const ov_model_t* model)`

File: docs/snippets/ov_dynamic_shapes.c (hunk 1 above)
Comment (zhaixuejun1993) [reply docs/snippets/ov_dynamic_shapes.c]: Following with C++, I think this part need to be changed after GPU enable.

File: docs/snippets/ov_dynamic_shapes.c (hunk 2 above)
Comment (zhaixuejun1993) [reply docs/snippets/ov_dynamic_shapes.c]: > We need free str_partial_shape here.

Done

File: docs/snippets/ov_dynamic_shapes.c (hunk 3 above)
Comment (zhaixuejun1993) [reply docs/snippets/ov_dynamic_shapes.c]: > For C sample, we need release all allocated objects explicitly - call ov_partial_shape_free to release partial_shape.

Done

File: docs/snippets/ov_dynamic_shapes.c (hunk 4 above)
Comment (zhaixuejun1993) [reply docs/snippets/ov_dynamic_shapes.c:86]: > Do clean up here for C sample.

Done

File: docs/snippets/ov_dynamic_shapes.c (hunk 5 above)
Comment (zhaixuejun1993) [reply docs/snippets/ov_dynamic_shapes.c:109]: In here, C++ check dynamic or not by `model->output(0).get_partial_shape().is_dynamic()`, which according to the partial shape. So using `ov_partial_shape_is_dynamic()` for C

**Review by ilyachur — APPROVED** [review]

**Review by riverlijunjie — APPROVED** [review]

**Comment by zhaixuejun1993** [comment]
Hi @ilyachur , all CI check had passed, pls help merge it, thanks.

**Pull request merged**